From 30756495c5c4c314c6cf7b45f52dc0da8c92c5e9 Mon Sep 17 00:00:00 2001
From: Rashmica Gupta <rashmica@linux.ibm.com>
Date: Tue, 25 Oct 2022 19:34:11 +1100
Subject: [PATCH] libpldm: Convert to the kernel MCTP socket interface
Signed-off-by: Rashmica Gupta <rashmica@linux.ibm.com>
Change-Id: Id34c393d6ea66ebd794e79b62df29f92662172e0
Patch Tracking Bug: b/277252180
Upstream-Status: Submitted
Upstream Info / review: https://gerrit.openbmc.org/c/openbmc/libpldm/+/58057
(Submitted by someone outside google)
Justification: Under review
---
 include/libpldm/mctp.h | 110 +++++++++++++++++++++++++++++++++++++++++
 src/requester/pldm.c   |   1 +
 2 files changed, 111 insertions(+)
 create mode 100644 include/libpldm/mctp.h
diff --git a/include/libpldm/mctp.h b/include/libpldm/mctp.h
new file mode 100644
index 0000000..271251c
--- /dev/null
+++ b/include/libpldm/mctp.h
@@ -0,0 +1,110 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * mctp: userspace utility for managing the kernel MCTP stack.
+ *
+ * Interim definitions for the MCTP kernel interface. While the MCTP kernel
+ * API is going through upstream, system toolchains may not yet have the
+ * required definitions for mctp tools. This header provides fallback
+ * implementations (protected by #ifdef guards) of those new interfaces.
+ *
+ * Copyright (c) 2021 Code Construct
+ * Copyright (c) 2021 Google
+ */
+
+#ifndef _MCTP_H
+#define _MCTP_H
+
+// #include "config.h"
+
+#if HAVE_LINUX_MCTP_H
+#include <linux/mctp.h>
+#endif
+#include <linux/if_link.h>
+
+#ifndef AF_MCTP
+#define AF_MCTP 45
+#endif
+
+/* MCTP serial line discipline */
+#ifndef N_MCTP
+#define N_MCTP 28
+#endif
+
+#include <stdint.h>
+
+#ifndef MCTP_NET_ANY
+
+typedef uint8_t mctp_eid_t;
+
+struct mctp_addr {
+	mctp_eid_t s_addr;
+};
+
+struct sockaddr_mctp {
+	unsigned short int smctp_family;
+	unsigned short __smctp_pad0;
+	int smctp_network;
+	struct mctp_addr smctp_addr;
+	uint8_t smctp_type;
+	uint8_t smctp_tag;
+	uint8_t __smctp_pad1;
+};
+
+#define MCTP_NET_ANY 0
+#define MCTP_ADDR_ANY 0xff
+#define MCTP_TAG_OWNER 0x08
+
+#endif /* MCTP_NET_ANY */
+
+#ifndef SIOCMCTPALLOCTAG
+#define SIOCMCTPALLOCTAG SIOCPROTOPRIVATE
+struct mctp_ioc_tag_ctl {
+	mctp_eid_t peer_addr;
+	__u8 tag;
+	__u16 flags;
+};
+
+#endif
+
+#ifndef MAX_ADDR_LEN
+#define MAX_ADDR_LEN 32
+#endif
+
+/* Added in v5.16 */
+#ifndef MCTP_OPT_ADDR_EXT
+
+#define MCTP_OPT_ADDR_EXT 1
+
+struct sockaddr_mctp_ext {
+	struct sockaddr_mctp smctp_base;
+	int smctp_ifindex;
+	uint8_t smctp_halen;
+	uint8_t __smctp_pad0[3];
+	unsigned char smctp_haddr[MAX_ADDR_LEN];
+};
+
+#endif /* MCTP_OPT_ADDR_EXT */
+
+#ifndef IFLA_MCTP_MAX
+
+/* From if_link.h */
+enum {
+	IFLA_MCTP_UNSPEC,
+	IFLA_MCTP_NET,
+	__IFLA_MCTP_MAX,
+};
+
+#define IFLA_MCTP_MAX (__IFLA_MCTP_MAX - 1)
+
+#endif /* IFLA_MCTP_MAX */
+
+/* setsockopt(2) options */
+#ifndef SOL_MCTP
+#define SOL_MCTP 285
+#endif
+
+#ifndef RTNLGRP_MCTP_IFADDR
+#define RTNLGRP_MCTP_IFADDR 34
+#endif
+
+#endif /* _MCTP_H */
diff --git a/src/requester/pldm.c b/src/requester/pldm.c
index 17d2492..b83fd96 100644
--- a/src/requester/pldm.c
+++ b/src/requester/pldm.c
@@ -1,4 +1,5 @@
 /* SPDX-License-Identifier: Apache-2.0 OR GPL-2.0-or-later */
+#include "libpldm/mctp.h"
 #include <libpldm/base.h>
 #include <libpldm/pldm.h>
 #include <libpldm/transport.h>
-- 
2.43.0.rc2.451.g8631bc7472-goog
